### PR TITLE
Allow setting clean options

### PR DESCRIPTION
**Pull request by harryadel**
https://github.com/Meteor-Community-Packages/meteor-collection2/issues/432

### CODE REVIEW

**Comment by harryadel** [comment]
I guess it's good to go @Prinzhorn. Any last comments?

File: README.md [hunk 1]
```diff
@@ -50,10 +51,6 @@ This package requires the [simpl-schema](https://github.com/aldeed/simple-schema
 - [Problems](#problems)
   - [SubObjects and Arrays of Objects](#subobjects-and-arrays-of-objects)
   - [Disable Displaying Collection Name in Error Message](#Disable-Displaying-Collection-Name-in-Error-Message)
-- [Contributing](#contributing)
```
Comment (Prinzhorn) [inline README.md:53]: The removal seems unrelated, but I guess that's on purpose?

File: README.md [hunk 2]
```diff
@@ -365,6 +362,23 @@ instance for a Mongo.Collection instance. For example:
...
+```js
+import Collection2 from 'meteor/aldeed:Collecion2'
+
+Collection2.cleanOptions = {
+        filter: false,
+        autoConvert: false,
+        removeEmptyStrings: false,
+        trimStrings: false,
+        removeNullsFromArrays: false,
+      }
```
Comment (Prinzhorn) [inline README.md:378]: Indentation looks wrong, but that might just be a different style

File: README.md [hunk 3]
```diff
@@ -365,6 +362,23 @@ instance for a Mongo.Collection instance. For example:
...
 ```
 
+## Schema Clean Options
+
+You can set clean options globally in collection2. 
+
+```js
+import Collection2 from 'meteor/aldeed:Collecion2'
+
+Collection2.cleanOptions = {
```
Comment (Prinzhorn) [inline README.md:372]: Maybe showcase the defaults here and add a comment

```js
// The values shown are the default options used internally. Overwrite them if needed.
Collection2.cleanOptions = {
  filter: true,
  autoConvert: true,
  removeEmptyStrings: true,
  trimStrings: true,
  removeNullsFromArrays: true,
}

// Or you can update individual options.
Collection2.cleanOptions.filter = false;
```

File: README.md [hunk 4]
```diff
@@ -365,6 +362,23 @@ instance for a Mongo.Collection instance. For example:
 MyCollection.simpleSchema().validate(doc);
 ```
 
+## Schema Clean Options
+
+You can set clean options globally in collection2.
```
Comment (Prinzhorn) [inline README.md:367]: Maybe like this: "You can set the `simpl-schema` clean options globally in collection2. They are merged with any options defined on the schema level."

File: tests/defaultCleanOptions.tests.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import expect from 'expect';
+import Collection2 from 'meteor/aldeed:collection2';
+
+describe('defaultCleanOptions', function () {
```
Comment (Prinzhorn) [inline tests/defaultCleanOptions.tests.js:4]: You've renamed it from ~`defaultCleanOptions`~ to `cleanOptions`, but maybe this was kept on purpose?

**Review by Prinzhorn — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (harryadel) [reply README.md:53]: Yes as these notes are already present in https://github.com/Meteor-Community-Packages/meteor-collection2/blob/master/CONTRIBUTING.md

**Comment by harryadel** [comment]
@Prinzhorn Kindly recheck.

**Review by Prinzhorn — APPROVED** [review]

**Comment by Prinzhorn** [comment]
Thanks!

**Comment by StorytellerCZ** [comment]
LGTM

**Comment by StorytellerCZ** [comment]
@harryadel would be nice if we could set this per collection or somehow disable it for a specific operation. Now looking at this I think I have a bug in one of my apps where this is trimming a place that should not be trimmed.

**Comment by Prinzhorn** [comment]
> would be nice if we could set this per collection or somehow disable it for a specific operation

Can't you do exactly that already?

Per collection: unless you are re-using a schema, you can pass `clean` to the `SimplSchema` constructor https://github.com/longshotlabs/simpl-schema#set-default-options-for-one-schema
Per operation: https://github.com/Meteor-Community-Packages/meteor-collection2#inserting-or-updating-without-cleaning

**Comment by harryadel** [comment]
@Prinzhorn You can add the updated version to your app now.

@StorytellerCZ huh, that's weird. Will you please open up an issue so we can better track this?

**Comment by Prinzhorn** [comment]
I just upgraded and unfortunately it does not work as expected and I wonder if it ever did.

Prior to this PR there was a const with the default clean options

https://github.com/Meteor-Community-Packages/meteor-collection2/blob/b0e856db590d78423fec33ed01021ada7404c84a/package/collection2/collection2.js#L18-L24

and for every clean action it was merged together with the clean option defined on the schema level plus the ones passed to this particular call

https://github.com/Meteor-Community-Packages/meteor-collection2/blob/b0e856db590d78423fec33ed01021ada7404c84a/package/collection2/collection2.js#L371-L382

I was assuming `schema._cleanOptions` could be unset, otherwise `(schema._cleanOptions || {})` would not make sense.

**However**, it is never undefined and simpl-schema already has a set of defaults:

https://github.com/longshotlabs/simpl-schema/blob/ea69d4eb7b4d8cc1151353989444a447bf0abc77/package/lib/SimpleSchema.js#L876-L888

https://github.com/longshotlabs/simpl-schema/blob/ea69d4eb7b4d8cc1151353989444a447bf0abc77/package/lib/SimpleSchema.js#L64-L67

What this means is that `defaultCleanOptions` (now `Collection2.cleanOptions`) never did anything. Maybe nobody noticed because the options are identical to the simpl-schema defaults anyway.

What that means is:

1. It can be entirely removed
2. What I was hoping for in #432 seems impossible given the way simpl-schema handles the clean options

Are my observation correct?

**Comment by harryadel** [comment]
@Prinzhorn Great write up and analysis.

Indeed since `...(schema._cleanOptions || {})` is set after `...defaultCleanOptions` it overrides any values you'd set in `defaultCleanOptions` rendering it meaningless. 

>What this means is that defaultCleanOptions (now Collection2.cleanOptions) never did anything. Maybe nobody noticed because the options are identical to the simpl-schema defaults anyway.

You're right, I guess it was set as precautionary measure in case simple schema values were not present for any reason.

>it can be entirely removed
What I was hoping for in Expose defaultCleanOptions #432 seems impossible given the way simpl-schema handles the clean options

I might release a fix/feature removing it altogether but I don't think we're in a hurry though.

>My use-case would be possible, however, via a different global setting that defaults to {} and is merged after schema._cleanOptions, but I think this entire thing is a mess and it should probably not be made more complex. In the end nobody will understand where the options actually come from.

We end up looping back to the original point. This should be handled within simpl-schema and non of collection2 concern. But hey, we learnt something along the way so thank you for that! <3

**Comment by harryadel** [comment]
>Indeed since ...(schema._cleanOptions || {}) is set after ...defaultCleanOptions it overrides any values you'd set in defaultCleanOptions rendering it meaningless.

This case should've accounted for in the tests and we should've listened for schema clean options instead of collection2 property.

https://github.com/Meteor-Community-Packages/meteor-collection2/blob/master/tests/defaultCleanOptions.tests.js#L15

**Comment by harryadel** [comment]
@Prinzhorn wouldn't this problem be solved if set `defaultCleanOptions` after `...(schema._cleanOptions || {})` but we then end up tampering with simpl-schema options. Honestly this whole entire debacle of how to simple schema clean options should work is out of collection2 goals.

**Comment by Prinzhorn** [comment]
> Honestly this whole entire debacle of how to simple schema clean options should work is out of collection2 goals.

I agree, that's what I meant with

> but I think this entire thing is a mess and it should probably not be made more complex. In the end nobody will understand where the options actually come from.

Originally I thought since collection2 already has `defaultCleanOptions` it would be a no-brainer to expose it. But now that I know more about the situation it's not worth the added complexity like you said.

**Pull request merged**